Refactor `IPCompleter` Matcher API

krassowski opened the pull request:
Closes #12820 and closes #13735. Supersedes #13734.

Partially fixes #13155 by exposing full context to matchers (but not the hooks yet - that would be a breaking change; I can follow up with it).

This is a gentle, backward and forward-compatible rewrite of Matcher API. I attempted to design the API for performance and future extensibility, while retaining compatibility with existing code.

There is quite some legacy logic in this codebase, which I tried to carefully preserve and document by making any such behaviour explicit, but in a few places I encountered behaviour which I don't know how to tackle (no unit tests, nor comments - not sure if existing behaviour was intended or incidental due to accumulated code complexity). I would highly appreciate help on such cases, suggestions on how to handle deprecations (if any are needed) and general feedback on this PR.

## User-facing changes

### Completion type is returned in the provisional API for most matchers

| Before | After |
|----|----|
| ![Screenshot from 2022-09-07 04-04-28](https://user-images.githubusercontent.com/5832902/188779612-0ff381e2-91db-497d-87f8-fb25d63d3d56.png) | ![Screenshot from 2022-09-07 04-04-32](https://user-images.githubusercontent.com/5832902/188779666-0f15c2b4-fd49-42b4-8800-0dd345137401.png) |

### Dictionary completions are restricted to dictionary keys

By default completions in dictionary context will only provide dictionary key matches. This is user-configurable.

The heuristic for this can be improved in the future iterations.

 Before | After |
|----|----|
![Screenshot from 2022-09-07 04-19-41](https://user-images.githubusercontent.com/5832902/188781277-b99e1104-c5dc-42ce-957e-ce100711074c.png) | ![Screenshot from 2022-09-07 04-18-00](https://user-images.githubusercontent.com/5832902/188781131-12ca3474-3578-41f6-b06a-c024923e5e1b.png) |

### Configuration

New configuration options were added:
- `IPCompleter.suppress_competing_matchers` allows to instruct which matchers should (or should not) short-circuit to return candidate completions ignoring other matchers
- `IPCompleter.disable_matchers` enables selective disabling of matchers

Existing configuration options are now aliases:
- `IPCompleter.merge_completions = False` is an alias for `IPCompleter.suppress_competing_matchers = True`

## Matcher API

Briefly, the new Matcher is defined as:

```python
MatcherAPIv1 = Callable[[str], list[str]]
MatcherAPIv2 = Callable[[CompletionContext], MatcherResult]

Matcher = Union[MatcherAPIv1, MatcherAPIv2]
```

where `MatcherAPIv1` is the same public API as exposed via (undocumented) `IPCompleter.custom_matchers` attribute since v8.0, https://github.com/ipython/ipython/pull/12130), `MatcherAPIv2` is the new API, and subsequent versions can be added in the future.

### Matcher API v2

The `MatcherResult` of API v2 is a dictionary, as and resolves an old TODO comment:

https://github.com/ipython/ipython/blob/7f51a0332edd0c675c2d314ca3e62df7ef041281/IPython/core/completer.py#L2142-L2145

```python
class MatcherResult(TypedDict):
    #: list of candidate completions
    completions: Union[Sequence[SimpleCompletion], Sequence[_JediCompletionLike]]   # simplified

    #: suffix of the provided ``CompletionContext.token``, if not given defaults to full token.
    matched_fragment: NotRequired[str]

    #: whether to suppress results from all other matchers (True), some
    #: matchers (set of identifiers) or none (False); default is False.
    suppress: NotRequired[Union[bool, Set[str]]]

    #: identifiers of matchers which should NOT be suppressed
    do_not_suppress: NotRequired[Set[str]]

    #: are completions already ordered and should be left as-is? default is False.
    ordered: NotRequired[bool]
```

This will enable adding additional attributes to the dictionary without breaking compatibility in the future. It also makes the API simple for users, as they only need to provide `{'completions': list[SimpleCompletion]}`.

`SimpleCompletion` is a container for completion metadata allowing to add type and in future additional attributes. After profiling initialisation, different forms of attribute/item access and memory use I chose to use a custom class with slots over `NamedTuple` or `TypedDict` due to inherent performance advantage. This is consistent with `Completion` class which also uses `__slots__`.

```python
class SimpleCompletion:
    __slots__ = ["text", "type"]

    text: str
    type: str
```

`CompletionContext` is intended to provide matchers with information they need to generate completion candidates:

```python
class CompletionContext(NamedTuple):
    token: str
    full_text: str
    cursor_position: int
    cursor_line: int

    @cached_property
    def text_until_cursor(self) -> str:
        ...

    @cached_property
    def line_with_cursor(self) -> str:
        ...
```

### Switching between APIs and Matcher configuration

A decorator was added enabling demarking API versions and configuring Matcher-level options. The decorator is optional for v1 API, but required to use v2 API

<details>

```python
def completion_matcher(
    *, priority: float = None, identifier: str = None, api_version=1
):
    """Adds attributes describing the matcher.
    Parameters
    ----------
    priority : Optional[float]
        The priority of the matcher, determines the order of execution of matchers.
        Higher priority means that the matcher will be executed first. Defaults to 50.
    identifier : Optional[str]
        identifier of the matcher allowing users to modify the behaviour via traitlets,
        and also used to for debugging (will be passed as ``origin`` with the completions).
        Defaults to matcher function ``__qualname__``.
    api_version: Optional[int]
        version of the Matcher API used by this matcher.
        Currently supported values are 1 and 2.
        Defaults to 1.
    """

    def wrapper(func: Matcher):
        func.matcher_priority = priority
        func.matcher_identifier = identifier or func.__qualname__
        func.matcher_api_version = api_version
        return func

    return wrapper
```

</details>

## Maintaining compatibility

Existing methods such as `dict_key_matches` were not touched; instead wrapper methods converting from API v1 to API v2 and adding desired metadata (when applicable) were added; those methods have a new suffix `_matcher`. This approach allows users who monkey-patch matcher v1 methods in `IPCompleter` to continue using their existing code without any changes.

## Follow-up tasks

To be discussed in separate issues and addressed in separate PRs:

- benchmark `type` as a string vs as an enum to inform whether it is worth switching to enum
- investigate how to integrate relevance score of sort (in `sortText ` in LSP terms)

krassowski committed:
Refactor `IPCompleter` Matcher API

krassowski committed:
Shim TypedDict and NotRequired at runtime until

the lowest supported Python version catches up.

krassowski committed:
Update ipdoctest test

krassowski commented:
Is there an automated way to update doctest with `ipdoctest`, or do I copy line by line? I added new configuration with verbose explanation and it just happens that `In [2]: %config IPCompleter` is used as an example in `IPython/core/magics/config.py`.

Carreau commented:
> Is there an automated way to update doctest with `ipdoctest`, or do I copy line by line? I added new configuration with verbose explanation and it just happens that `In [2]: %config IPCompleter` is used as an example in `IPython/core/magics/config.py`.

No there is no automatic ways. I copy and past also with light editting.

krassowski committed:
Implement `priority`, `do_not_suppress`, add tests and docs.

krassowski committed:
Correct suppression defaults, add a test for #13735

krassowski committed:
Merge branch 'main' into completion-matcher

joelostblom commented:
Awesome! The key-only completion for dictionaries will be super convenient!

krassowski committed:
Improve type hinting and documentation

krassowski commented:
@meeseeksdev tag tab-completion

krassowski committed:
Move `typing_extensions` optional dependency

krassowski committed:
`highlight` → `code-block`

krassowski commented:
Ok, I would say this is good for a review and hopefully merge.

I am happy to commit to maintaining this and fixing any problems that might emerge as a result of merging this (which I don't expect frankly since everything should be 100% backward compatible - but we know how it is with big projects).

Documentation-wise I hit a minor hiccup with documenting the `TypedDict` but I submitted a bugfix to sphinx autodoc which solves the issue https://github.com/sphinx-doc/sphinx/pull/10806 and all members will show up once that is accepted and released.

krassowski committed:
Fix backslash combining matchers (they require `text_until_cursor`).

krassowski committed:
Merge branch 'main' into completion-matcher

krassowski committed:
Relax constraint on limit to allow no limit

krassowski commented on IPython/core/completer.py:
```diff
@@ -417,6 +537,189 @@ def __hash__(self):
+    #: Cursor line in ``full_text``.
+    cursor_line: int
+
+    #: The maximum number of completions that will be used downstream.
+    #: Matchers can use this information to abort early.
+    #: The built-in Jedi matcher is currently excepted from this limit.
+    # If not given, return all possible completions.
+    limit: Optional[int]
```
For most scenarios we could remove the `limit` to reduce the API and instead allow `MatcherResult` to accept generators, but there are some scenarios where knowing how much work needs to be done is beneficial (anything where pre-allocation may be required).

I am happy to remove `limit` and instead allow generators in `MatcherResult` if that would make merging this PR easier :)

Carreau commented on IPython/core/completer.py:
```diff
@@ -417,6 +537,189 @@ def __hash__(self):
+    @property
+    @lru_cache(maxsize=None)  # TODO change to @cache after dropping Python 3.7
+    def line_with_cursor(self) -> str:
+        return self.full_text.split("\n")[self.cursor_line]
```
We are already 3.8+ https://github.com/ipython/ipython/blob/main/setup.cfg#L29, I think you mean after dropping Python 3.8. 

One easy thing is at the top of the file you can do 

```
if sys.version_info < (3, 8):
   from functools import lru_cahce
   cache = lru_cache(maxsize=None)
else:
   from functools import cache
```

I grep from `sys_version_info` every time we drop a Python version.

krassowski replied:
Yes, it should have been < 3.8. In ab60d3197788b943db0cf81d9279c28e52ec9225 I switched to `@cached_property` instead as using `@property @cache` as it was a premature optimisation (it would be more memory efficient as per documentation, but at any time there is only one instance of `CompletionContext`, so there is no need to worry about key-sharing optimisation - and for the same reason we can switch it to `dataclass` which has a benefit of narrower API)

Carreau commented on IPython/core/completer.py:
```diff
@@ -417,6 +537,189 @@ def __hash__(self):
+
+
+@sphinx_options(show_inherited_members=True, exclude_inherited_from=["dict"])
+class SimpleMatcherResult(_MatcherResultBase, TypedDict):
```
DO you need to inherit both if `_MatcherResultBase` is already TypedDict. 
and do we want to TypedDict, or simply make a datalass with all the possible existing fields, even if it's slightly incompatible ?

krassowski replied:
> DO you need to inherit both if _MatcherResultBase is already TypedDict.

No, not at runtime, but yes if we want to have nice documentation (https://github.com/sphinx-doc/sphinx/pull/10806); a quirk of how `TypedDict` it was implemented. I added a comment in the body on this.

> and do we want to TypedDict, or simply make a datalass with all the possible existing fields, even if it's slightly backward-incompatible ?

I hesitated here and considered making it a dataclass to, but I preferred `TypedDict` for two reasons:
- in future releases we can add a new `NotRequired` field without breaking compatibility (and give notice if we want to make it required in the further future)
- it does not require users to import any specific class from IPython to wrap the results (or create a custom class complying with the API), so it would keep it simple (while argument name and type safety is provided by typecheckers)

Carreau commented on IPython/utils/docs.py:
```diff
@@ -0,0 +1,7 @@
+# encoding: utf-8
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
```
```suggestion
```

When present, those things are just relegate of the Python 2 and pre-github era, there is no need to add them to new files. We just avoid removing to not pollute history.

Carreau commented on docs/sphinxext/apigen.py:
```diff
@@ -24,14 +24,9 @@
-    def __init__(self, **kwargs):
-        for k, v in kwargs.items():
-            setattr(self, k, v)
-                                        
```
+1

Carreau commented:
Thanks, I tried to do a quick pass, and will try to get a deeper look when on my way to CZI meeting next week. 

In general I think you can be a little bit less careful on backward compatibility, and if you find really old things try to remove them if unnecessary, we only have a few dependees that use old APIs.

krassowski commented on IPython/core/completer.py:
```diff
@@ -417,6 +537,189 @@ def __hash__(self):
+        self.type = type
+
+    def __repr__(self):
+        return f"<SimpleCompletion text={self.text!r} type={self.type!r}>"
```
I decided not to use a dataclass due to a performance overhead.

krassowski committed:
Remove outdated header as suggested

Co-authored-by: Matthias Bussonnier <bussonniermatthias@gmail.com>

krassowski committed:
Switch `CompletionContext` to `dataclass`, use `cached_property`

Data class has narrower API than named tuple (does not expose item getters)
which means it is less likely that the downstream code would be broken in
the future if we introduce new attributes. Also, compared to other places
where memory-efficient named tuple is used, `CompletionContext` is only
created once per completion and thus does not require as low overhead.
For the same reason we can forgo slightly more memory-efficient
`@property @cache` stack and just use `@cached_property`.

Carreau commented:
Ok, let's just try this and refine if needs there is.

jdtsmith commented:
Playing with the new completer in an Emacs mode I'm building.  So far so good; great work.  Seems faster as well.  

I had formerly overridden the `matchers` property to remove file and magic matches in certain contexts.  `disable_matchers` is obviously the new way to do that.  But is the `matchers` property no longer consulted for the list of enabled matchers?  If so, its value is probably misleading.

krassowski commented:
Thank you for the feedback @jdtsmith! Would you mind opening an issue with a minimal reproducible example? I don't see what could be the issue as:
- `IPCompleter.matchers` was and remained a dynamically generated list defined as a property (the behaviour here did not change)
- `Completer.custom_matchers` are still prepended to the matchers list (above) and used extensively in tests, so I don't believe there is a regression here
Thanks!

And yes, this PR lays a foundation for improved performance (see https://github.com/ipython/ipython/issues/13752 for an example).
